### PR TITLE
WIP: Port `crates#index` to use Diesel

### PR DESCRIPTION
**Pull request by sgrif**
All the tests are failing since they set up their test data using the
rust-postgres connection, and I still need to port them over. I'm done
for the evening though, and I doubt porting the tests will change the
code significantly, so I figured I'd toss this up now to see if there's
any feedback.

### CODE REVIEW

File: migrations/20170307211844_versions_yanked_is_not_nullalbe/up.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE versions ALTER COLUMN yanked SET NOT NULL;
```
Comment (carols10cents) [inline migrations/20170307211844_versions_yanked_is_not_nullalbe/up.sql]: 👍 👍 👍 👍

File: src/app.rs [hunk 2]
```diff
@@ -54,7 +54,8 @@ impl App {
             .helper_threads(if config.env == ::Env::Production {3} else {1})
             .build();
         let diesel_db_config = r2d2::Config::builder()
-            .pool_size(if config.env == ::Env::Production {1} else {1})
+            .pool_size(if config.env == ::Env::Production {50} else {1})
```
Comment (carols10cents) [inline src/app.rs]: This 50 is going to have to come down as we discovered yesterday

File: src/category.rs [hunk 3]
```diff
@@ -21,6 +24,15 @@ pub struct Category {
     pub crates_cnt: i32,
 }
 
+#[derive(Associations)]
+#[belongs_to(Category)]
+#[table_name="crates_categories"]
+#[allow(dead_code)] // FIXME: Hey @sgrif add better many-to-many support to Diesel, wouldya?
```
Comment (carols10cents) [inline src/category.rs]: ❤️

File: src/app.rs (hunk 2 above)
Comment (sgrif) [inline src/app.rs]: I thought prod had 120 connections available? I told it to not try to keep more than 5 idles at a time so it should never get bad on staging.

Comment (carols10cents) [inline src/app.rs]: Oh I see. I thought this would mean I couldn't deploy it to staging.

Comment (carols10cents) [inline src/app.rs]: Wait... `min_idle` means "not more than 5 idles"? That sounds like `max_idles` to me...?

Comment (sgrif) [inline src/app.rs]: Yeah, r2d2's naming/behavior/defaults are all really questionable here.

Comment (sgrif) [inline src/app.rs]: You could totally take down staging by sending 20 concurrent requests though, but I figured that's not a huge concern.

File: migrations/20170307211844_versions_yanked_is_not_nullalbe/up.sql (hunk 1 above)
Comment (sgrif) [inline migrations/20170307211844_versions_yanked_is_not_nullalbe/up.sql]: inb4 prod has a ton of rows where this is null

**Comment by sgrif** [comment]
This requires porting more endpoints than I thought. Going after some other endpoints for now. No point leaving this open.

**Pull request closed without merging**